spec/strings_file_validation_helper_spec.rb

```diff
@@ -0,0 +1,66 @@
+require 'spec_helper'
+
+module Fastlane
+  UI = FastlaneCore::UI unless Fastlane.const_defined?('UI')
+
+  module Helper
+    class StringsFileValidationHelper
+      # Inspects the given `.strings` file for duplicated keys.
+      #
+      # @param [String] file The path to the file to inspect.
+      def self.find_duplicated_keys(file:)
+        found_keys = {}
+        duplicated_keys = []
+        File.read(file).lines.each_with_index do |line, index|
+          key = extract_key(line: line)
+
+          next if key.nil?
+
+          # Check if that key was already encountered in the past
+          existing = found_keys[key]
+
+          duplicated_keys.append(key) unless existing.nil?
+          # TODO: Use proper Fastlane UI methods for this, or better rely on
+          # the consumer to read the return Hash (itself a TODO) and print an
+          # error. This is an helper, after all.
+          puts "warning: key `#{key}` on line #{index + 1} is a duplicate of a similar key found on line #{existing + 1}" unless existing.nil?
+
+          # Memorize the line at which this key appeared in the source `.strings`
+          # file
+          found_keys[key] = index
+        end
+
+        duplicated_keys
+      end
+
+      def self.extract_key(line:)
+        regexp = /\s*"([^"]+.*)"\s*=/
```

Comment:
We discussed the fact that, when it comes to `.strings`, parsing is tough and it might very well be one of those case where you try to solve a problem with a RegExp and you're now left with two problems.

But, I'm not sure how else to go about finding duplicated keys otherwise.

I'd say that since the purpose of this RegExp is quite limited, and we can "test the heck out of it", it's an okay approach in this case.

You might recognize the code around here. I lifted it from your original suggestion at paaHJt-2S5-p2#comment-5716. However, I modified the RegExp to avoid the strings substitution, which I think makes the code clearer.

Reply:
Yeah, tough call there.

Notice that for example that means that the action won't work on `InfoPlist.strings` file which typically does not use `""` quotes around key names… not sure if that was one of the potential file we'd want to target with this action, but worth noting anyway.

It also won't work with keys that contain `\"` in their name — because of the `[^"]+` thing. The `.*` that follows might accidentally workaround this, actually rendering the preceding `[^"]+` kinda useless, but that also risks making the `.*` be greedy (e.g. `"Login to a \"%@\" account" = "Login to a \"%@\" account"; // Here %@ = the service name, like Google`)

While we'd hope in an ideal world that we'd try to avoid such keys, and despite the fact that our new tooling will now accept keys that are not the English copy (`AppLocalizedString("key", value: "English copy", comment: "blah")`) which would even lessen the risk of having such cases, we know that those cases currently exist in the current codebase and `Localizable.strings` files…

I don't have a magic solution for this though, tbh. Maybe using a Parser Combinator or at least a parsing simple state machine would make it easier than using a RegEx when it comes to handling backslash escapes and quote balancing? 🤔  Though maybe not, and maybe that'd also be killing a fly with a bazooka…